src/runtime_src/doc/toc/xrt_native_apis.rst

```diff
@@ -0,0 +1,618 @@
+
+
+
+Kernel and Run
+--------------
+
+The XRT kernel APIs support creating of kernel handle (or object in C++) from currently loaded xclbin.  The kernel handle is used to execute the kernel function on the hardware instance (Compute Unit or CU) of the kernel.  
+
+A Run handle/object represents an execution of the kernel. Upon finishing the kernel execution, the Run handle/object can be reused to invoke the same kernel function if desired. 
+
+The following topics are discussed below
+
+       - Obtaining kernel handle/object from XCLBIN
+       - Getting the bank group index of a kernel argument
+       - Reading and write CU mapped registers
+       - Execution of kernel and dealing with the associated run
+       - Other kernel execution related API
+       
+
+Obtaining kernel handle/object from XCLBIN
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The kernel handle (or object) is created from the device, XCLBIN UUID and the kernel name. 
+
+.. code:: c
+      :number-lines: 35
+           
+           xuid_t xclbin_uuid;
+           xrtXclbinGetUUID(xclbin,xclbin_uuid);
+
+           xrtKernelHandle kernel = xrtPLKernelOpen(device, xclbin_uuid, "kernel_name");
+           ....
+           ....
+           xrtKernelClose(kernel);
+
+
+In the above code example
+ 
+      - The UUID of the XCLBIN is retrieved by the API ``xrtXclbinGetUUID`` 
+      - The kernel is created by the API ``xrtPLKernelOpen``
+      - The kernel is closed by the API ``xrtKernelClose``
+
+**Note**: For the kernel with more than 1 CU, a kernel handle (or object) should represent all the CUs having identical interface connectivity. If all the CUs of the kernel are not having identical connectivity, the specific CU name(s) should be used to obtain a kernel handle (or object) to represent the subset of CUs with identical connectivity. Otherwise XRT will do this selection internally to select a group of CUs and discard the rest of the CUs (discarded CUs are not used during the execution of a kernel).  
+
+As an example, assume a kernel name is foo having 3 CUs foo_1, foo_2, foo_3. The CUs foo_1 and foo_2 are connected to DDR bank 0, but the CU foo_3 is connected to DDR bank 1. 
+
+       - Opening kernel handle for foo_1 and foo_2 (as they have identical interface connection)
+       
+         .. code:: c++
+               :number-lines: 35
+                  
+                    cu_group_1 = xrtPLKernelOpen(device, xclbin_uuid, "foo:{foo_1,foo_2}");     
+   
+       - Opening kernel handle for foo_3
+          
+         .. code:: c++
+               :number-lines: 35
+                  
+                    cu_group_2 = xrtPLKernelOpen(device, xclbin_uuid, "foo:{foo_3}");     
+
+
+
+**C++**: In C++, xrt::kernel object can be created from the constructor of ``xrt::kernel`` class. 
+
+.. code:: c++
+      :number-lines: 35
+          
+           auto xclbin_uuid = device.load_xclbin("kernel.xclbin");
```

Comment:
What is the type of a `xclbin_uuid`? What is the life-time of the magical resources behind the scene?

Reply:
In C++ everything is managed.  Why do you worry about lifetime, are you ever concerned bout lifetime of std::vector and std::blah.  Unless otherwise stated all everything in C++ is managed via scoping.

Reply:
no action needed for this PR related to this comment

Reply:
> In C++ everything is managed. Why do you worry about lifetime, are you ever concerned bout lifetime of std::vector and std::blah. Unless otherwise stated all everything in C++ is managed via scoping.

I am not worried by life-time in ISO C++ because it is done by a standardization committee and explained by an official document. ;-)